proxy/proxy.go

```diff
@@ -57,56 +58,94 @@ type BaseURLResolver interface {
 // 	- logging errors and proxy request timing to stdout
 //
 // Note that this will panic if `resolver` is nil.
-func NewHandlerFunc(timeout time.Duration, resolver BaseURLResolver) http.HandlerFunc {
+func NewHandlerFunc(config types.FaaSConfig, resolver BaseURLResolver) http.HandlerFunc {
 	if resolver == nil {
 		panic("NewHandlerFunc: empty proxy handler resolver, cannot be nil")
 	}
 
-	proxyClient := http.Client{
+	proxyClient := NewProxyClientFromConfig(config)
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		if r.Body != nil {
+			defer r.Body.Close()
+		}
+
+		switch r.Method {
+		case http.MethodPost,
+			http.MethodPut,
+			http.MethodPatch,
+			http.MethodDelete,
+			http.MethodGet:
+
+			proxyRequest(w, r, proxyClient, resolver)
+
+		default:
+			w.WriteHeader(http.StatusMethodNotAllowed)
+		}
+	}
+}
+
+// NewProxyClientFromConfig creates a new http.Client designed for proxying requests and enforcing
+// certain minimum configuration values.
+func NewProxyClientFromConfig(config types.FaaSConfig) *http.Client {
+	maxIdleConns := config.MaxIdleConns
+	if maxIdleConns < 1 {
+		maxIdleConns = 1024
+	}
+
+	maxIdleConnsPerHost := config.MaxIdleConnsPerHost
+	if maxIdleConnsPerHost < 1 {
+		maxIdleConnsPerHost = 1024
+	}
+
+	timeout := config.ReadTimeout
+	if timeout <= 0*time.Second {
+		timeout = 10 * time.Second
+	}
+
+	return NewProxyClient(timeout, maxIdleConns, maxIdleConnsPerHost)
+}
+
+// NewProxyClient creates a new http.Client designed for proxying requests, this is exposed as a
+// convenience method for internal or advanced uses. Most people should use NewProxyClientFromConfig.
+func NewProxyClient(timeout time.Duration, maxIdleConns int, maxIdleConnsPerHost int) *http.Client {
+	return &http.Client{
 		// these Transport values ensure that the http Client will eventually timeout and prevents
 		// infinite retries. The default http.Client configure these timeouts.  The specific
 		// values tuned via performance testing/benchmarking
 		//
 		// Additional context can be found at
 		// - https://medium.com/@nate510/don-t-use-go-s-default-http-client-4804cb19f779
 		// - https://blog.cloudflare.com/the-complete-guide-to-golang-net-http-timeouts/
+		//
+		// Additionally, these overrides for the default client enable re-use of connections and prevent
+		// CoreDNS from rate limiting under high traffic
+		//
+		// See also two similar projects where this value was updated:
+		// https://github.com/prometheus/prometheus/pull/3592
+		// https://github.com/minio/minio/pull/5860
 		Transport: &http.Transport{
 			Proxy: http.ProxyFromEnvironment,
 			DialContext: (&net.Dialer{
 				Timeout:   timeout,
 				KeepAlive: 1 * time.Second,
+				DualStack: true,
```

Comment:
Why do we have DualStack as `true` now?

Reply:
I copied it from gateway

Reply:
OK 👍

This is now part of the DefaultTransport, seems fine to me.

https://golang.org/pkg/net/http/#RoundTripper